src/solvers/sat/satcheck_ipasir.cpp

```diff
@@ -0,0 +1,194 @@
+/*******************************************************************\
+
+Module: External SAT Solver Binding
+
+Author: Norbert Manthey, nmanthey@amazon.com
+
+\*******************************************************************/
+
+#ifndef _MSC_VER
+#include <inttypes.h>
+#endif
+
+#include <cassert>
+#include <stack>
+
+#include <util/threeval.h>
+
+#include "satcheck_ipasir.h"
+
+#ifdef HAVE_IPASIR
+
+extern "C"
+{
+#include <ipasir.h>
+}
+
+/*
+
+Interface description:
+https://github.com/biotomas/ipasir/blob/master/ipasir.h
+
+Representation:
+Variables for a formula start with 1! 0 is used as termination symbol.
+
+*/
+
+tvt satcheck_ipasirt::l_get(literalt a) const
+{
+  if(a.is_true())
+    return tvt(true);
+  else if(a.is_false())
+    return tvt(false);
+
+  tvt result;
+
+  // compare to internal no_variables number
+  if(a.var_no()>=(unsigned)no_variables())
+    return tvt::unknown();
+
+  const int val=ipasir_val(solver, a.var_no());
+
+  if(val>0)
+    result=tvt(true);
+  else if(val<0)
+    result=tvt(false);
+  else
+    return tvt::unknown();
+
+  if(a.sign())
+    result=!result;
+
+  return result;
+}
+
+const std::string satcheck_ipasirt::solver_text()
+{
+  return std::string(ipasir_signature());
+}
+
+void satcheck_ipasirt::lcnf(const bvt &bv)
+{
+  forall_literals(it, bv)
+  {
+    if(it->is_true())
+      return;
+    else if(!it->is_false())
+      INVARIANT(it->var_no()<(unsigned)no_variables(),
+             "reject out of bound variables");
+  }
+
+  forall_literals(it, bv)
+  {
+    if(!it->is_false())
+    {
+      // add literal with correct sign
+      ipasir_add(solver, it->sign()?-it->var_no():it->var_no());
+    }
+  }
+  ipasir_add(solver, 0); // terminate clause
+
+  clause_counter++;
+}
+
+propt::resultt satcheck_ipasirt::prop_solve()
+{
+  INVARIANT(status!=statust::ERROR, "there cannot be an error");
+
+  {
+    messaget::status() <<
+      (no_variables()-1) << " variables, " <<
+      clause_counter << " clauses" << eom;
+  }
+
+  // use the internal representation, as ipasir does not support reporting the
+  // status
+  if(status==statust::UNSAT)
+  {
+    messaget::status() <<
+      "SAT checker inconsistent: instance is UNSATISFIABLE" << eom;
+  }
+  else
+  {
+    // if assumptions contains false, we need this to be UNSAT
+    bvt::const_iterator it = std::find_if(assumptions.begin(), assumptions.end(), is_false);
+    const bool has_false = it != assumptions.end();
+
+    if(has_false)
+    {
+      messaget::status() <<
+        "got FALSE as assumption: instance is UNSATISFIABLE" << eom;
+    }
+    else
+    {
+      forall_literals(it, assumptions)
+        if(!it->is_false())
+          ipasir_assume(solver, it->sign()?-it->var_no():it->var_no());
+
+      // solve the formula, and handle the return code (10=SAT, 20=UNSAT)
+      int solver_state==ipasir_solve(solver);
+      if(10==solver_state)
```

Comment:
I am surprised that the == parses?

Reply:
That slipped through copy and paste, and the compiler was happy ... if fixed it and tested another round with the LIBSOLVER build locally.